nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.io.InputStreamCallback;
+import org.apache.nifi.processors.pulsar.AbstractPulsarProducerProcessor;
+import org.apache.nifi.schema.access.SchemaNotFoundException;
+import org.apache.nifi.serialization.MalformedRecordException;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.RecordReaderFactory;
+import org.apache.nifi.serialization.RecordSetWriter;
+import org.apache.nifi.serialization.RecordSetWriterFactory;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSchema;
+import org.apache.nifi.serialization.record.RecordSet;
+import org.apache.nifi.stream.io.StreamUtils;
+import org.apache.nifi.util.StringUtils;
+import org.apache.pulsar.client.api.Producer;
+
+@Tags({"Apache", "Pulsar", "Record", "csv", "json", "avro", "logs", "Put", "Send", "Message", "PubSub", "1.0"})
+@CapabilityDescription("Sends the contents of a FlowFile as individual records to Apache Pulsar using the Pulsar 1.x client API. "
+    + "The contents of the FlowFile are expected to be record-oriented data that can be read by the configured Record Reader. "
+    + "The complementary NiFi processor for fetching messages is ConsumePulsarRecord.")
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@WritesAttribute(attribute = "msg.count", description = "The number of messages that were sent to Pulsar for this FlowFile. This attribute is added only to "
+        + "FlowFiles that are routed to success.")
+@SeeAlso({PublishPulsar.class, ConsumePulsar.class, ConsumePulsarRecord.class})
+public class PublishPulsarRecord extends AbstractPulsarProducerProcessor<byte[]> {
+
+    private static final List<PropertyDescriptor> PROPERTIES;
+
+    static {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.addAll(AbstractPulsarProducerProcessor.PROPERTIES);
+        properties.add(RECORD_READER);
+        properties.add(RECORD_WRITER);
+        PROPERTIES = Collections.unmodifiableList(properties);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final String topic = context.getProperty(TOPIC).evaluateAttributeExpressions(flowFile).getValue();
+
+        if (StringUtils.isBlank(topic)) {
+            getLogger().error("Invalid topic specified {}", new Object[] {topic});
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        // Read the contents of the FlowFile into a byte array
+        final byte[] messageContent = new byte[(int) flowFile.getSize()];
```

Comment:
For Record-oriented processors we can't read the contents into an array - they are intended to be stream-oriented and avoid buffering too much into memory. We may well get multi-gigabyte FlowFIles.